shared/coordinator/src/coordinator.rs

```diff
@@ -965,39 +997,106 @@ impl<T: NodeIdentity> Coordinator<T> {
     ) -> std::result::Result<TickResult, CoordinatorError> {
-        if self.check_timeout(unix_timestamp, self.config.round_witness_time) {
-            // TODO: Punish idle witnesses
-            self.epoch_state.first_round = false.into();
-            self.progress.step += 1;
+        if !self.check_timeout(unix_timestamp, self.config.round_witness_time) {
+            return Ok(TickResult::Ticked);
+        };
+        // Round witness time is over, we can now start the next round
 
-            let current_round = self.current_round_unchecked();
-            let height = current_round.height;
-            let num_witnesses = current_round.witnesses.len() as u16;
-            self.move_clients_to_exited(height);
+        // TODO: Punish idle witnesses
+        self.epoch_state.first_round = false.into();
+        self.progress.step += 1;
+
+        let current_round = self.current_round_unchecked();
+        let height = current_round.height;
+        let num_witnesses = current_round.witnesses.len() as u16;
+
+        let mut removed_clients_idx = self.move_clients_to_exited(height);
+
+        // If there are not witnesses, then we can't distinguish from
+        // the situation where only witness nodes disconnected or everyone
+        // disconnected. We just set everyone to withdrawn state and change
+        // to Cooldown.
+        if num_witnesses == 0 {
+            self.withdraw_all()?;
+            self.start_cooldown(unix_timestamp);
+            return Ok(TickResult::Ticked);
+        }
 
-            // If there are not witnesses, then we can't distinguish from
-            // the situation where only witness nodes disconnected or everyone
-            // disconnected. We just set everyone to withdrawn state and change
-            // to Cooldown.
-            if num_witnesses == 0 {
-                self.withdraw_all()?;
-                self.start_cooldown(unix_timestamp);
-                return Ok(TickResult::Ticked);
+        let witnesses = &self.current_round_unchecked().witnesses;
+        let mut client_times_last_index_updated = 0;
+        if let Some(mut agreed_values) =
+            self.get_consensus_on_reported_times(witnesses, 1.0 / 3.0, 5.0)
+        {
+            // If we have exited clients, we need to remove their times from the agreed times as
+            // well, as to keep consistency with the indices.
+            // We want to remove from the last index to the first, to avoid shuffling the array.
+            removed_clients_idx.sort_by(|c1: &_, c2| c2.cmp(c1));
+            for i in removed_clients_idx.iter() {
+                agreed_values.remove(*i);
+            }
+            msg!(
+                "[tick_round_witness] Agreed time values: {:?}",
+                agreed_values
+            );
+
+            // Here we assume the offset is the same for all of the witnesses, we should run a consensus on that.
+            let start_offset_in_global_times = witnesses[0].training_times.offset;
```

Comment:
just as another argument for centrally managing the training times offset, the 0th witness could have lied with the offset sent and this would lead to errors in this code